runtime/pkg/middleware/usage.go

```diff
@@ -0,0 +1,61 @@
+package middleware
+
+import (
+	"context"
+	"time"
+
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/rilldata/rill/runtime/pkg/usage"
+	"github.com/rilldata/rill/runtime/server/auth"
+	"google.golang.org/grpc"
+)
+
+func UsageStreamServerInterceptor() grpc.StreamServerInterceptor {
+	return func(
+		srv interface{},
+		ss grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler,
+	) error {
+		subject := auth.GetClaims(ss.Context()).Subject()
+		if subject == "" {
+			subject = "anonymous"
+		}
+
+		newCtx := usage.ContextWithUsageDims(ss.Context(), *usage.String("user_id", subject))
+		wss := grpc_middleware.WrapServerStream(ss)
+		wss.WrappedContext = newCtx
+
+		start := time.Now()
+		defer func() {
+			// Emit usage metric
+			usage.GetClient().Emit(newCtx, "request/time", float64(time.Since(start).Milliseconds()))
```

Comment:
does it also contain the request name and any other dimensions ?

Reply:
request method + other dimensions created based on request params